File: DecentralisedPID/DynamixelPID.cpp
#include <array>
#include <cmath>
#include <algorithm>

// Define constants
const int PWM_SCALE = 884;
const double MAX_PWM = 1.0;

// Define types for clarity
using Vector2 = std::array<double, 2>;
using Vector4 = std::array<double, 4>;
//
//  State vectors shoudl be of form
//      [ Theta1Target]
//      [ Theta2Target]
//      [  Theta1Arm  ]
//      [  Theta2Arm  ]
//
//    MATLAB CODE HAS BETTER COMMENTING PLEASE REFER TO THAT    
//
// Function implementing the PID controller
void DynamixelPID(const Vector4& CurrentState,
                  Vector4& PastState,
                  double TimeSinceLast,
                  Vector2& RunningErrorSums,
                  Vector2& PWMValues) {
    // PID gain matrices (identity matrices in this case)
    const double K_p[2][2] = {{1.0, 0.0}, {0.0, 1.0}};
    const double K_I[2][2] = {{1.0, 0.0}, {0.0, 1.0}};
    const double K_D[2][2] = {{1.0, 0.0}, {0.0, 1.0}};

    // Calculate current error
    Vector2 CurrentError = {
        CurrentState[0] - CurrentState[2],
        CurrentState[1] - CurrentState[3]
    };

    // Calculate error derivative
    Vector2 ErrorDeriv = {
        ((CurrentState[0] - CurrentState[2]) - (PastState[0] - PastState[2])) / TimeSinceLast,
        ((CurrentState[1] - CurrentState[3]) - (PastState[1] - PastState[3])) / TimeSinceLast
    };

    // Update running error sums
    RunningErrorSums[0] += CurrentError[0] * TimeSinceLast;
    RunningErrorSums[1] += CurrentError[1] * TimeSinceLast;

    // Compute PID output
    Vector2 PWMOutput = {
        K_p[0][0] * CurrentError[0] + K_D[0][0] * ErrorDeriv[0] + K_I[0][0] * RunningErrorSums[0],
        K_p[1][1] * CurrentError[1] + K_D[1][1] * ErrorDeriv[1] + K_I[1][1] * RunningErrorSums[1]
    };

    // Clamp PWM values to [-1, 1]
    for (int i = 0; i < 2; ++i) {
        if (std::abs(PWMOutput[i]) > MAX_PWM) {
            PWMOutput[i] = (PWMOutput[i] > 0 ? 1 : -1);
        }
    }

    // Scale to Dynamixel range and round
    PWMValues[0] = std::round(PWMOutput[0] * PWM_SCALE);
    PWMValues[1] = std::round(PWMOutput[1] * PWM_SCALE);

    // Update past state
    PastState = CurrentState;
}

File: DecentralisedPID/DynamixelPID.m
function [PWMValues, PastState, RunningErrorSums] = DynamixelPID(CurrentState,PastState, TimeSinceLast, RunningErrorSums)
%takes State vectors from current and last known state of 4x1 form 
%       [ Theta1Target ]
%  S  = [ Theta2Target ]
%       [ Theta1Arm    ] 
%       [ Theta2Arm    ]
%  as well as time since last state in seconds, and running sum of errors
%  in 2x1 vector form
%  
%  E  = [RunningErrorSum1]
%       [RunningErrorSum2] 
%
%  Outputs column vector of PWM values(from -884 to 884, corresponding to -100% to 100%),
%  as well as updating the "PastState" vector and "RunningError"
%==========================================================================
%Set K values for P I and D (change these as needed to tune)
K_p=[1 0; 0 1];
K_I=[1 0; 0 1];
K_D=[1 0; 0 1];

%calculate the error of current state
CurrentError= [CurrentState(1)-CurrentState(3);...
               CurrentState(2)-CurrentState(4) ];

%calculate error deriv
ErrorDeriv= [(CurrentState(1)-CurrentState(3))-(PastState(1)-PastState(3));...
             (CurrentState(2)-CurrentState(4))-(PastState(2)-PastState(4))]...
             /TimeSinceLast;

%calculate running error
RunningErrorSums=RunningErrorSums+(CurrentError*TimeSinceLast);

%create PWM initial Values
PWMValues=(K_p*CurrentError)+(K_D*ErrorDeriv)+(K_I*RunningErrorSums);

%make sure PWM values dont exceed -100% to 100%
if abs(PWMValues(1))>1
    PWMValues(1)=PWMValues(1)/abs(PWMValues(1));
end
if abs(PWMValues(2))>1
    PWMValues(2)=PWMValues(2)/abs(PWMValues(2));
end 

%Scale PWM values to the range of Dynamixel and make sure theyre intergers
PWMValues=round(PWMValues*884);

%Update the PastState
PastState=CurrentState;

end
